### PR TITLE
[ADD] sale_order_product_recommendation_secondary_unit: New module

### PR DESCRIPTION
**Pull request by chienandalu**
This module adds secondary unit to the product recommendation wizard.

cc @Tecnativa TT19379

### CODE REVIEW

**Comment by rafaelbn** [comment]
Hello @chienandalu , thanks for this contribution!

Travis is :red_circle: 

```
2019-10-28 15:37:46,290 5769 ERROR openerp_test odoo.addons.sale_order_product_recommendation.tests.test_recommendation: ERROR: test_transfer (odoo.addons.sale_order_product_recommendation.tests.test_recommendation.RecommendationCase)

2019-10-28 15:37:46,290 5769 ERROR openerp_test odoo.addons.sale_order_product_recommendation.tests.test_recommendation: ` Products get transferred to SO.

2019-10-28 15:37:46,290 5769 ERROR openerp_test odoo.addons.sale_order_product_recommendation.tests.test_recommendation: Traceback (most recent call last):

2019-10-28 15:37:46,290 5769 ERROR openerp_test odoo.addons.sale_order_product_recommendation.tests.test_recommendation: `   File "/home/travis/build/OCA/sale-workflow/sale_order_product_recommendation/tests/test_recommendation.py", line 139, in test_transfer

2019-10-28 15:37:46,290 5769 ERROR openerp_test odoo.addons.sale_order_product_recommendation.tests.test_recommendation: `     wizard.action_accept()

2019-10-28 15:37:46,291 5769 ERROR openerp_test odoo.addons.sale_order_product_recommendation.tests.test_recommendation: `   File "/home/travis/build/OCA/sale-workflow/sale_order_product_recommendation/wizards/sale_order_recommendation.py", line 134, in action_accept

2019-10-28 15:37:46,291 5769 ERROR openerp_test odoo.addons.sale_order_product_recommendation.tests.test_recommendation: `     wiz_line._prepare_update_so_line())

2019-10-28 15:37:46,291 5769 ERROR openerp_test odoo.addons.sale_order_product_recommendation.tests.test_recommendation: `   File "/home/travis/OCB-11.0/odoo/models.py", line 4614, in update

2019-10-28 15:37:46,291 5769 ERROR openerp_test odoo.addons.sale_order_product_recommendation.tests.test_recommendation: `     record[name] = value

2019-10-28 15:37:46,291 5769 ERROR openerp_test odoo.addons.sale_order_product_recommendation.tests.test_recommendation: `   File "/home/travis/OCB-11.0/odoo/models.py", line 4818, in __setitem__

2019-10-28 15:37:46,291 5769 ERROR openerp_test odoo.addons.sale_order_product_recommendation.tests.test_recommendation: `     return self._fields[key].__set__(self, value)

2019-10-28 15:37:46,291 5769 ERROR openerp_test odoo.addons.sale_order_product_recommendation.tests.test_recommendation: ` KeyError: 'product_qty'

2019-10-28 15:37:46,291 5769 INFO openerp_test odoo.addons.sale_order_product_recommendation.tests.test_recommendation: Ran 3 tests in 3.526s

2019-10-28 15:37:46,292 5769 ERROR openerp_test odoo.addons.sale_order_product_recommendation.tests.test_recommendation: FAILED

2019-10-28 15:37:46,292 5769 INFO openerp_test odoo.addons.sale_order_product_recommendation.tests.test_recommendation:  (errors=1)

2019-10-28 15:37:46,292 5769 ERROR openerp_test odoo.modules.module: Module sale_order_product_recommendation: 0 failures, 1 errors
```

**Comment by rafaelbn** [comment]
@chienandalu I cannot test it the is not option to use secondary unit

File: sale_order_product_recommendation_secondary_unit/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    "license": "AGPL-3",
+    "application": False,
+    "installable": True,
+    "depends": [
+        "sale_order_product_recommendation",
+        "sale_order_secondary_unit",
+    ],
+    "data": [
+        "wizards/sale_order_recommendation_view.xml",
+    ],
```
Comment (sergio-teruel) [inline sale_order_product_recommendation_secondary_unit/__manifest__.py:20]: Add auto_install

File: sale_order_product_recommendation/wizards/sale_order_recommendation.py [hunk 2]
```diff
@@ -207,3 +203,19 @@ def _compute_price_unit(self):
     @api.onchange("units_included")
     def _onchange_units_included(self):
         self.is_modified = bool(self.sale_line_id or self.units_included)
+
+    @api.multi
```
Comment (sergio-teruel) [inline sale_order_product_recommendation/wizards/sale_order_recommendation.py]: remove multi

File: sale_order_product_recommendation/wizards/sale_order_recommendation.py [hunk 3]
```diff
@@ -207,3 +203,19 @@ def _compute_price_unit(self):
...
+    def _prepare_update_so_line(self):
+        """So we can extend PO update"""
+        return {
+            'product_qty': self.units_included,
+        }
+
+    @api.multi
```
Comment (sergio-teruel) [inline sale_order_product_recommendation/wizards/sale_order_recommendation.py]: remove multi

File: sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    @api.model
+    def _prepare_wizard_line(self, vals, order_line=False):
+        res = super()._prepare_wizard_line(vals, order_line)
+        secondary_uom_id = (
+            order_line and order_line.secondary_uom_id or
+            vals.get('product_id') and
+            vals['product_id'].sale_secondary_uom_id)
+        secondary_uom_qty = False
+        if not order_line and secondary_uom_id:
+            secondary_uom_qty = ceil(
```
Comment (sergio-teruel) [inline sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py]: Why do you use ceil??

File: sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        qty = float_round(
+            self.units_included / (factor or 1.0),
+            precision_rounding=self.secondary_uom_id.uom_id.rounding)
+        if float_compare(
+                self.secondary_uom_qty,
+                qty,
+                precision_rounding=self.secondary_uom_id.uom_id.rounding) != 0:
+            self.secondary_uom_qty = qty
+
+    @api.multi
```
Comment (sergio-teruel) [inline sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py]: remove multi

**Review by sergio-teruel — CHANGES_REQUESTED** [review]

**Comment by sergio-teruel** [comment]
@chienandalu I do not know if I'm doing well.
I sell a product with secondary units
Selección_445
I confirm the order and validate the picking
Selección_446
I create other sale order and clink on "recomended products".
I don't see secondary units and I can not write secondary unit fields
Selección_447

**Comment by chienandalu** [comment]
@sergio-teruel You should set the default secondary unit for sale for that product:
Selección_068
Selección_069

File: sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py (hunk 4 above)
Comment (chienandalu) [reply sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py]: To obtain a round unit of sencondary uom qty. For example, a case where de secondary uom is box of 12, and we get the recommendation to sell 3.2 of such unit, the conversion would be: 4 boxes of 12 units

**Comment by chienandalu** [comment]
> @chienandalu I cannot test it the is not option to use secondary unit

@rafaelbn Ok, maybe a better explanation in the readme is needed (or at least refer to the proper main module setup)

File: sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py (hunk 4 above)
Comment (sergio-teruel) [reply sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py]: But in sale order do not work in same way...

**Review by rafaelbn — COMMENTED** [review]
Waiting for a better README in order to test it. Thanks!

**Comment by chienandalu** [comment]
@sergio-teruel Check changes, please
@rafaelbn I extended the readme and hope you can test now

File: sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+    def _prepare_wizard_line(self, vals, order_line=False):
+        res = super()._prepare_wizard_line(vals, order_line)
+        secondary_uom_id = (
+            order_line and order_line.secondary_uom_id or
+            vals.get('product_id') and
+            vals['product_id'].sale_secondary_uom_id)
+        secondary_uom_qty = False
+        if not order_line and secondary_uom_id:
+            secondary_uom_qty = (
+                res.get('units_included', 0) / secondary_uom_id.factor)
```
Comment (sergio-teruel) [inline sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py]: (secondary_uom_id.factor or 1.0)

**Review by sergio-teruel — CHANGES_REQUESTED** [review]

File: sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py (hunk 6 above)
Comment (chienandalu) [reply sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py]: Looks like a constraint should be applied to that field, isn't it? I mean, how the factor could be 0?

Comment (chienandalu) [reply sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py]: I mean, in the same way regular uoms do: https://github.com/odoo/odoo/blob/11.0/addons/product/models/product_uom.py#L42-L45

Comment (sergio-teruel) [reply sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py]: But this factor is this (I think)
https://github.com/OCA/product-attribute/blob/11.0/product_secondary_unit/models/product_second_unit.py#L25

Comment (chienandalu) [reply sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py]: Changed

Comment (sergio-teruel) [reply sale_order_product_recommendation_secondary_unit/wizards/sale_order_recommendation.py]: oks...

**Review by sergio-teruel — APPROVED** [review]

**Review by pedrobaeza — APPROVED** [review]

**Pull request merged**